### DISCUSSION TITLE
Bintray (including JCenter) Sunsetting on May 1st, 2021 - Publish to Maven Central instead

### DISCUSSION THREAD

**Discussion by jhg023** [discord-jda/JDA#1538] [General]
Bintray (including JCenter) will be sunsetting on May 1st, 2021, meaning that it will no longer be available as an artifact repository. For more information, see: [here](https://jfrog.com/blog/into-the-sunset-bintray-jcenter-gocenter-and-chartcenter/).

The JDA artifact(s) will need to be hosted somewhere else, to which I recommend Maven Central. For more information on publishing to Maven Central, see this [YouTube video](https://www.youtube.com/watch?v=bxP9IuJbcDQ).

**Comment by MinnDevelopment** [comment]
We are already aware of this and looking into alternatives.

**Comment by Wild1145** [comment]
Is there any update on when a new home for this is expected? Github security alerts are now warning that we're pointing to a vulnerable endpoint because the service is being sun-setted so we're looking at options ASAP To move the repo on our end to somewhere new, but just wanted to know if that was already under way on your end?

**Reply by MinnDevelopment** [reply]
We are still working on it. You can use [jitpack](https://jitpack.com/#DV8FromTheWorld/JDA) in the meantime if you need an immediate alternative.

**Reply by Wild1145** [reply]
Thanks, greatly appreciated. My first thought was Jitpack but that doesn't seem to work at the moment for some reason either, looks like it tries to pull in a dependency that Jitpack can't build. 

`Failed to collect dependencies at com.github.DV8FromTheWorld:JDA:jar:v4.2.0 -> club.minnced:opus-java:pom:1.0.4: Failed to read artifact descriptor for club.minnced:opus-java:pom:1.0.4: Could not transfer artifact club.minnced:opus-java:pom:1.0.4 from/to jitpack.io (https://jitpack.io): Not authorized , ReasonPhrase:Unauthorized. -> [Help 1]`

If you've got any thoughts on why it's doing that / how to fix it until a more permanent solution is in place would be greatly appreciated!

**Reply by MinnDevelopment** [reply]
You can just exclude opus-java if you don't need audio. Otherwise, that dependency is loaded through jcenter.

**Reply by DV8FromTheWorld** [reply]
To be clear, `JCenter` is being sunset, but the packages _currently available_ on `JCenter` will be available through February of 2022 so none of your builds are going to break until Feb of 2022. 
Publishing new versions of packages is being cut off on March 31 2021, so after that point no additional version of JDA will be pushed to JCenter.
[Details available here](https://jfrog.com/blog/into-the-sunset-bintray-jcenter-gocenter-and-chartcenter/)

We are testing our a few options internally to decide what path to take for new packages. Version `4.2.1` will be published to the new repository once decided.

**Reply by Wild1145** [reply]
Appreciate the packages aren't going away in the immediate future, the reason it's been noticed (At least in my case) is Github's CodeQL have now declared anything on bintray.com to be a security risk under CWE-1104 which is why we're trying to move where we pull it from because all our builds now fail due to CodeQL Failing on PR's and any commits we now make to our repo's using JDA as an upstream dependency. 

The check from CodeQL can be found here - https://github.com/github/codeql/blob/main/java/ql/src/Security/CWE/CWE-1104/MavenPomDependsOnBintray.ql

**Reply by DV8FromTheWorld** [reply]
This seems kinda silly to deal with so soon given the actual 2022 deadline. Can you not disable this specific CodeQL check for now?

**Reply by Wild1145** [reply]
Not as far as I'm aware, we can mark it as a false positive / similar, but that only obviously is in the local repo. We can obviously ignore the error (Which is what we've done so far) but it is something we're looking on our projects to resolve ASAP to keep the number of security alerts to a minimum especially for something that in theory from our side should be as simple as changing a URL or just deleting an endpoint depending on where the artefacts get hosted in the future.

**Reply by DV8FromTheWorld** [reply]
Alrighty. I appreciate the detail and information. This is something on our immediate radar and is a blocker for releasing `4.2.1`, so rest assured it will be handled soon.

**Comment by Gamebuster19901** [comment]
If you're using gradle, you can use source dependencies to retrieve JDA directly from github. 

There is documentation of gradle source dependencies at: 
https://blog.gradle.org/introducing-source-dependencies

And a working example when using JDA as a dependency:
https://github.com/Gamebuster19901/ExciteBot/commit/1cba6d5ffed0efe108be0698a2fd1cc72af1e858

**Reply by DV8FromTheWorld** [reply]
Please note that this above example will be pulling from the release tag [v4.2.0](https://github.com/DV8FromTheWorld/JDA/releases/tag/v4.2.0) which was release on **June 26, 2020**.

If you are going to do this, you may want to pull from a commit hash instead.

**Reply by Gamebuster19901** [reply]
I believe it actually pulls from this commit: https://github.com/DV8FromTheWorld/JDA/commit/3a9fcd0ecc88f988ac81eb2c18527283a83f12d1 which was released 3 days later.

I don't believe you can specify a specific commit using gradle source dependencies. If you wish to use a specific commit, you can fork the repository, and rollback/fast forward/merge a branch to whatever commit you want.

Then in settings.gradle, you would change the username of the repository to your username:

```
sourceControl {
	gitRepository("https://github.com/USERNAME/JDA.git") {
		producesModule("net.dv8tion:JDA")
	}
}
```

And in build.gradle, pull from your branch:
```
    implementation('net.dv8tion:JDA') {
        version {
            branch = 'YOURBRANCH'
        }
    }
```

**Reply by DV8FromTheWorld** [reply]
Seems like a lot. I'd just use `jitpack` to be honest.

Either way, whether you're using `jitpack` or gradle's `gitRespository`, you're still going to be relying on `JCenter` as the dependencies defined by JDA are being pulled from `JCenter` in JDA's `build.gradle.kts`.

**Comment by DV8FromTheWorld** [comment]
This had been addressed. See the current README details for information about how to get builds from our new repository location
https://github.com/dv8fromtheworld/jda#download

**Reply by greg6775** [reply]
Why not publish to Maven Central instead?

**Reply by DV8FromTheWorld** [reply]
Eventually we will, I didn't have the time to jump through the hoops of getting Maven Central setup. Additionally, releasing to Maven Central takes a while before things actually show up in Central which we also didn't want to deal with currently.

Eventually we will be double publishing.
